File: problems/java/MeetingScheduler/src/models/Meeting.java
package models;

import enums.RSVPStatus;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

public class Meeting {

    private final int id;
    private final String subject;
    private final User organizer;
    private final Map<User, RSVPStatus> participantStatus;
    private final MeetingRoom bookedRoom;
    private final Interval bookedInterval;
    private boolean active;
    private static final AtomicInteger nextId = new AtomicInteger(1);

    public Meeting(String subject, User organizer, List<User> participants, MeetingRoom room, Interval interval) {
        this.id = nextId.getAndIncrement();
        this.subject = subject;
        this.organizer = organizer;
        bookedRoom = room;
        bookedInterval = interval;
        active = true;
        participantStatus = new ConcurrentHashMap<>();

        for (User participant : participants) {
            participantStatus.put(participant, RSVPStatus.PENDING);
        }
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public void addParticipant(User participant) {
      participantStatus.putIfAbsent(participant, RSVPStatus.PENDING);
    }

    public void removeParticipant(User participant) {
        participantStatus.remove(participant);
    }

    public void updateParticipantStatus(User participant, RSVPStatus status) {
        participantStatus.computeIfPresent(participant, (k, v) -> status);
    }

    public List<User> getAllParticipants() {
        return new ArrayList<>(participantStatus.keySet());
    }

    public int getId() {
        return id;
    }

    public User getOrganizer() {
        return organizer;
    }

    public String getSubject() {
        return subject;
    }

    public MeetingRoom getBookedRoom() {
        return bookedRoom;
    }

    public Interval getBookedInterval() {
        return bookedInterval;
    }

}

File: problems/python/MeetingScheduler/src/models/meeting.py
from itertools import count
from threading import Lock

from enums.rsvp_status import RSVPStatus
from models.interval import Interval
from models.meeting_room import MeetingRoom


class Meeting:
    _next_id = count(1)
    _id_lock = Lock()

    def __init__(
        self,
        subject: str,
        organizer: "User",
        participants: list["User"],
        room: MeetingRoom,
        interval: Interval,
    ) -> None:
        with Meeting._id_lock:
            self.id = next(Meeting._next_id)
        self.subject = subject
        self.organizer = organizer
        self.booked_room = room
        self.booked_interval = interval
        self.active = True
        self._participant_status: dict["User", RSVPStatus] = {
            participant: RSVPStatus.PENDING for participant in participants
        }
        self._status_lock = Lock()

    def add_participant(self, participant: "User") -> None:
        with self._status_lock:
            self._participant_status.setdefault(participant, RSVPStatus.PENDING)

    def remove_participant(self, participant: "User") -> None:
        with self._status_lock:
            self._participant_status.pop(participant, None)

    def update_participant_status(self, participant: "User", status: RSVPStatus) -> None:
        with self._status_lock:
            if participant in self._participant_status:
                self._participant_status[participant] = status

    def get_all_participants(self) -> list["User"]:
        with self._status_lock:
            return list(self._participant_status.keys())
